ui-driven/ui-driven.go

```diff
@@ -0,0 +1,182 @@
+package uidriven
+
+import (
+	"time"
+
+	"github.com/temporalio/samples-go/ui-driven/proxy"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/sdk/workflow"
+)
+
+const (
+	RegisterStage = "register"
+	SizeStage     = "size"
+	ColorStage    = "color"
+	CompleteStage = "complete"
+)
+
+var (
+	TShirtSizes = []string{
+		"small",
+		"medium",
+		"large",
+	}
+
+	TShirtColors = []string{
+		"red",
+		"blue",
+		"black",
+	}
+)
+
+type TShirtOrder struct {
+	Email string
+	Size  string
+	Color string
+}
+
+type OrderStatus struct {
+	OrderID string
+	Stage   string
+}
+
+// Workflow is a workflow driven by interaction from a UI.
+func OrderWorkflow(ctx workflow.Context) error {
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 5 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	order := TShirtOrder{}
+
+	// Loop until we receive a valid email
+	for {
+		id, _, email := proxy.ReceiveRequest(ctx)
+
+		err := workflow.ExecuteActivity(ctx, RegisterEmail, email).Get(ctx, nil)
+		if err != nil {
+			err := proxy.SendErrorResponse(ctx, id, err)
+			if err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		order.Email = email
+
+		err = proxy.SendResponse(ctx, id, SizeStage, "")
+		if err != nil {
+			return err
+		}
+
+		break
+	}
+
+	// Loop until we receive a valid size
+	for {
+		id, _, size := proxy.ReceiveRequest(ctx)
+
+		err := workflow.ExecuteActivity(ctx, ValidateSize, size).Get(ctx, nil)
+		if err != nil {
+			err := proxy.SendErrorResponse(ctx, id, err)
+			if err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		order.Size = size
+
+		err = proxy.SendResponse(ctx, id, ColorStage, "")
+		if err != nil {
+			return err
+		}
+
+		break
+	}
+
+	// Loop until we receive a valid color
+	for {
+		id, _, color := proxy.ReceiveRequest(ctx)
+
+		err := workflow.ExecuteActivity(ctx, ValidateColor, color).Get(ctx, nil)
+		if err != nil {
+			err := proxy.SendErrorResponse(ctx, id, err)
+			if err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		order.Color = color
+
+		// Tell the UI the order is complete
+		err = proxy.SendResponse(ctx, id, CompleteStage, "")
+		if err != nil {
+			return err
+		}
+
+		break
+	}
+
+	err := workflow.ExecuteActivity(ctx, ProcessOrder, order).Get(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func StartOrderWorkflow(ctx workflow.Context, email string) (OrderStatus, error) {
```

Comment:
As we discussed on Slack, I think it would simplify the sample if we'd remove this overhead workflow and execute the OrderWorkflow with the e-mail as a parameter directly from the ui.

The ui then could simply query and wait for the workflow to get to the SizeStage.

The current approach:
![current](https://user-images.githubusercontent.com/5324856/120772979-25944a00-c521-11eb-84b2-96aaa187b4b5.png)


<details>
  <summary>Click to expand for current uml</summary>
  
  ```
  @startuml
ui -> temporal : ExecuteWorkflow(StartOrder, email)
temporal -> StartOrderWorkflow : Schedule()
StartOrderWorkflow -> temporal : ExecuteChildWorkflow(OrderWorkflow)
temporal -> OrderWorkflow : Schedule()

StartOrderWorkflow -> OrderWorkflow : SendRequest(RegisterStage, email)
OrderWorkflow -> StartOrderWorkflow : ReceiveResponse(nextStage)

ui -> temporal : ExecuteWorkflow(UpdateOrder, size)
temporal -> UpdateOrderWorkflow1 : Schedule()
UpdateOrderWorkflow1 -> OrderWorkflow : SignalExternalWorkflow(size)
OrderWorkflow --> UpdateOrderWorkflow1 : SignalExternalWorkflow(colorStage)
UpdateOrderWorkflow1 -> ui : UpdateOrderWorkflow1.Result()

ui -> temporal : ExecuteWorkflow(UpdateOrder, color)
temporal -> UpdateOrderWorkflow2 : Schedule()
UpdateOrderWorkflow2 -> OrderWorkflow : SignalExternalWorkflow(color)
OrderWorkflow --> UpdateOrderWorkflow2 : SignalExternalWorkflow(CompleteStage)
UpdateOrderWorkflow2 -> ui : UpdateOrderWorkflow2.Result()
@enduml
```

</details>

Without the StartOrder workflow:
![simpflified](https://user-images.githubusercontent.com/5324856/120772989-27f6a400-c521-11eb-870c-2d802dbd5291.png)


<details>
  <summary>Click to expand for uml</summary>
  
  ```
@startuml
ui -> temporal : ExecuteWorkflow(OrderWorkflow, email)
temporal -> OrderWorkflow : Schedule()

ui <- temporal : QueryWorkflow(Orderworkflow, currentStage)
note over ui : wait for Order to get to sizeStage

ui -> temporal : ExecuteWorkflow(UpdateOrder, size)
temporal -> UpdateOrderWorkflow1 : Schedule()
UpdateOrderWorkflow1 -> OrderWorkflow : SignalExternalWorkflow(size)
OrderWorkflow --> UpdateOrderWorkflow1 : SignalExternalWorkflow(colorStage)
UpdateOrderWorkflow1 -> ui : UpdateOrderWorkflow1.Result()

ui -> temporal : ExecuteWorkflow(UpdateOrder, color)
temporal -> UpdateOrderWorkflow2 : Schedule()
UpdateOrderWorkflow2 -> OrderWorkflow : SignalExternalWorkflow(color)
OrderWorkflow --> UpdateOrderWorkflow2 : SignalExternalWorkflow(CompleteStage)
UpdateOrderWorkflow2 -> ui : UpdateOrderWorkflow2.Result()
@enduml
```

</details>

Reply:
Simplified by starting the workflow from the UI as an empty order, then adding details via UpdateOrder. This means we only have one proxy workflow and no need of childworkflows with abandon policy.

Reply:
As discussed on Slack I've resisted using queries as that means polling which feels a bit ironic given the synchronous focused use case I'm trying to present :)

Reply:
I agree, that makes perfect sense. I think it's more readable now 👍 

Here's the uml to go with that:

<details>
  <summary>Click to expand</summary>

```
@startuml
ui -> temporal : ExecuteWorkflow(OrderWorkflow)
temporal -> OrderWorkflow : Schedule()

ui -> temporal : ExecuteWorkflow(UpdateOrder, RegisterStage, email)
temporal -> UpdateOrderWorkflow1 : Schedule()
UpdateOrderWorkflow1 -> OrderWorkflow : SignalExternalWorkflow(email)
OrderWorkflow --> UpdateOrderWorkflow1 : SignalExternalWorkflow(SizeStage)
UpdateOrderWorkflow1 -> ui : UpdateOrderWorkflow1.Result()

ui -> temporal : ExecuteWorkflow(UpdateOrder, SizeStage, size)
temporal -> UpdateOrderWorkflow2 : Schedule()
UpdateOrderWorkflow2 -> OrderWorkflow : SignalExternalWorkflow(size)
OrderWorkflow --> UpdateOrderWorkflow1 : SignalExternalWorkflow(ColorStage)
UpdateOrderWorkflow2 -> ui : UpdateOrderWorkflow2.Result()

ui -> temporal : ExecuteWorkflow(UpdateOrder, ColorStage, color)
temporal -> UpdateOrderWorkflow3 : Schedule()
UpdateOrderWorkflow3 -> OrderWorkflow : SignalExternalWorkflow(color)
OrderWorkflow --> UpdateOrderWorkflow3 : SignalExternalWorkflow(CompleteStage)
UpdateOrderWorkflow3 -> ui : UpdateOrderWorkflow3.Result()
@enduml

```

</details>


![ui_driven](https://user-images.githubusercontent.com/5324856/120782789-a4da4b80-c52a-11eb-822b-d3609b6cbff9.png)

Reply:
@nadilas Please can you update the graph for the shipping changes? Thanks :)